homeassistant/components/shelly/switch.py

```diff
@@ -120,46 +127,17 @@ def async_setup_block_entry(
     coordinator = config_entry.runtime_data.block
     assert coordinator
 
-    # Add Shelly Motion as a switch
-    if coordinator.model in MOTION_MODELS:
-        async_setup_entry_attribute_entities(
-            hass,
-            config_entry,
-            async_add_entities,
-            {("sensor", "motionActive"): MOTION_SWITCH},
-            BlockSleepingMotionSwitch,
-        )
-        return
-
-    if config_entry.data[CONF_SLEEP_PERIOD]:
-        return
-
-    # In roller mode the relay blocks exist but do not contain required info
-    if (
-        coordinator.model in [MODEL_2, MODEL_25]
-        and coordinator.device.settings["mode"] != "relay"
-    ):
-        return
-
-    relay_blocks = []
-    assert coordinator.device.blocks
-    for block in coordinator.device.blocks:
-        if block.type != "relay" or (
-            block.channel is not None
-            and is_block_channel_type_light(
-                coordinator.device.settings, int(block.channel)
-            )
-        ):
-            continue
-
-        relay_blocks.append(block)
-        unique_id = f"{coordinator.mac}-{block.type}_{block.channel}"
-        async_remove_shelly_entity(hass, "light", unique_id)
-
-    if not relay_blocks:
-        return
+    async_setup_entry_attribute_entities(
+        hass, config_entry, async_add_entities, BLOCK_RELAY_SWITCHES, BlockRelaySwitch
+    )
 
-    async_add_entities(BlockRelaySwitch(coordinator, block) for block in relay_blocks)
+    async_setup_entry_attribute_entities(
+        hass,
+        config_entry,
+        async_add_entities,
+        BLOCK_SLEEPING_MOTION_SWITCH,
+        BlockSleepingMotionSwitch,
+    )
```

Comment:
Its unfortunate that `async_setup_entry_attribute_entities` ends up calling `async_add_entites` multiple times since each call makes a new task... but this is an existing design problem